bite-size-notes/week-5-c-Reachability-Analysis.md

```diff
@@ -0,0 +1,215 @@
+
+<img width="805" alt="Reachability" src="https://github.com/UW-CTRL/AA548-spr2024/assets/160373451/9a103a0b-d195-4cf0-bf9e-08ea850e22df">
+
+On the figure above you see T is the Target set. Inside target set h(x) is negative, and outside of the set h(x) is positive. Our starting point is $x(t_0)$ and any point on the trajectory is defined as $x(t)$. h function takes the state X and gives a single number R, it is a scalar valued function. We are trying to minimize the cost function h.
+
+<img width="864" alt="Screen Shot 2024-04-29 at 23 53 07" src="https://github.com/kutaydemiralay/AA548-spr2024/assets/160373451/ea162ed5-7012-4308-8910-1fa06eb89572">
+
+We may also care about entering Target set not on final time but any time between [0,T]
+## Hamilton-Jacobi Reachability
+Hamilton-Jacobi Reachability is a specific method used in the context of reachability analysis. It uses the Hamilton-Jacobi-Bellman (HJB) or Hamilton-Jacobi- Isaacs (HJI) equations, which are partial differential equations that describe the evolution of the reachability boundary with time. 
+
+There is a HJ toolbox written in in stanford ASL written for python, and matlab available, and can be imported with the command: import hj_reachability in phython. It is a toolbow that allows us to solve reachability problems.
+
+-Example From HJ toolbox
+
+-Example System:Air3d  (Pursuer trying to reach to an Evader)
+
+-System Dynamics:
+
+$$
+\begin{aligned}
+\dot{x} & = v \cos(\theta), \\
+\dot{y} & = v \sin(\theta), \\
+\dot{\theta} & = \omega.
+\end{aligned}
+$$
+
+### Code:
+
+```python
+import hj # Assuming 'hj' is a hypothetical module for Hamilton-Jacobi analysis
+
+# Initialize the dynamics for a 3D air system
+dynamics = hj.systems.Air3d()
+
+# Create a grid with specific boundary conditions and lattice parameters
+grid = hj.Grid.from_lattice_parameters_and_boundary_conditions(
+    hj.sets.Box(np.array([-6., -10., 0.]), np.array([20., 10., 2 * np.pi])),
+    (51, 40, 50),
+    periodic_dims=2
+)
+
+# Compute the values as the norm of states, adjusted by a radius of 5 units
+values = jnp.linalg.norm(grid.states[..., :2], axis=-1) - 5
+
+# Define solver settings with very high accuracy and a specific postprocessor for reachability
+solver_settings = hj.SolverSettings.with_accuracy(
+    "very_high",
+    hamiltonian_postprocessor=hj.solver.backwards_reachable_tube
+)
+time = 0.
+target_time = -2.8
+target_values = hj.step(solver_settings, dynamics, grid, time, values, target_time)
+```
+
+-"dynamics" line helps setting up dynamics
+
+-"grid"  line specifiies mesh, the grid points we are solving the ODE over
+
+-"vaules" line specifies the terminal point (we need a termenial point to sole ODE backwards in time. And specifies the minimum distance allowable between 2 planes, which is 5, so we form a circle of 5.
```

Comment:
There seems to be a typo in the variable name 'values'. also to make the variables name more readable and pretty you can put them between `this` character: ``